matchms/similarity/spectrum_similarity_functions.py

```diff
@@ -5,69 +5,111 @@
 
-    for idx in range(len(spec1)):
-        intensity = spec1[idx, 1]
-        mz = spec1[idx, 0]
-        matches = numpy.where((numpy.abs(spec2[:, 0] - spec1[idx, 0] + shift) <= tolerance))[0]
-        for match in matches:
-            power_prod_spec1 = ((mz ** mz_power) * (intensity ** intensity_power))
-            power_prod_spec2 = ((spec2[match][0] ** mz_power) * (spec2[match][1] ** intensity_power))
-            matching_pairs.append((idx, match, power_prod_spec1 * power_prod_spec2))
 
-    return matching_pairs
+@numba.njit
+def find_matches(spec1, spec2, tolerance, shift):
+    """Faster search for matching peaks.
+    Makes use of the fact that spec1 and spec2 contain ordered peak m/z (from
+    low to high m/z).
+
+    Parameters
+    ----------
+    spec1:
+        Spectrum peaks and intensities as numpy array. Peak mz values must be ordered.
+    spec2:
+        Spectrum peaks and intensities as numpy array. Peak mz values must be ordered.
+    tolerance
+        Peaks will be considered a match when <= tolerance appart.
+    shift
+        Shift spectra peaks by shift. The default is 0.
+
+    Returns
+    -------
+    matches
+        List containing entries of type (idx1, idx2).
+
+    """
+    lowest_idx = 0
+    matches = []
+    for peak1_idx in range(spec1.shape[0]):
+        mz = spec1[peak1_idx, 0]
+        low_bound = mz - tolerance
+        high_bound = mz + tolerance
+        for peak2_idx in range(lowest_idx, spec2.shape[0]):
+            mz2 = spec2[peak2_idx, 0] + shift
+            if mz2 > high_bound:
+                break
+            if mz2 < low_bound:
+                lowest_idx = peak2_idx
+            else:
+                matches.append((peak1_idx, peak2_idx))
+    return matches
 
 
 def get_peaks_array(spectrum: SpectrumType) -> numpy.ndarray:
     """Get peaks mz and intensities as numpy array."""
     return numpy.vstack((spectrum.peaks.mz, spectrum.peaks.intensities)).T
 
 
-def score_best_matches(matching_pairs: list, spec1: numpy.ndarray,
+@numba.njit(fastmath=True)
+def score_best_matches(matching_pairs: numpy.ndarray, spec1: numpy.ndarray,
                        spec2: numpy.ndarray, mz_power: float = 0.0,
                        intensity_power: float = 1.0) -> Tuple[float, int]:
     """Calculate cosine-like score by multiplying matches. Does require a sorted
     list of matching peaks (sorted by intensity product)."""
+    score = float(0.0)
+    used_matches = int(0)
     used1 = set()
     used2 = set()
-    score = 0.0
-    used_matches = []
-    for match in matching_pairs:
-        if not match[0] in used1 and not match[1] in used2:
-            score += match[2]
-            used1.add(match[0])  # Every peak can only be paired once
-            used2.add(match[1])  # Every peak can only be paired once
-            used_matches.append(match)
+    for i in range(matching_pairs.shape[0]):
+        if not matching_pairs[i, 0] in used1 and not matching_pairs[i, 1] in used2:
```

Comment:
In tests `if` was never false. Can a test be written for that case?

Reply:
Same here, that case will occur in the tests.